Mathlib/RingTheory/AlgebraicIndependent/TranscendenceBasis.lean

```diff
@@ -98,6 +114,98 @@ theorem IsTranscendenceBasis.isAlgebraic [Nontrivial R] (hx : IsTranscendenceBas
+  simpa using hax
+
+theorem isTranscendenceBasis_iff_algebraicIndependent_isAlgebraic [Nontrivial R] :
+    IsTranscendenceBasis R x ↔
+      AlgebraicIndependent R x ∧ Algebra.IsAlgebraic (adjoin R (range x)) A :=
+  ⟨fun h ↦ ⟨h.1, h.1.isTranscendenceBasis_iff_isAlgebraic.mp h⟩,
+    fun ⟨ind, alg⟩ ↦ ind.isTranscendenceBasis_iff_isAlgebraic.mpr alg⟩
+
+theorem IsTranscendenceBasis.sumElim_comp [NoZeroDivisors A] {x : ι → S} {y : ι' → A}
+    (hx : IsTranscendenceBasis R x) (hy : IsTranscendenceBasis S y) :
+    IsTranscendenceBasis R (Sum.elim y (algebraMap S A ∘ x)) := by
+  cases subsingleton_or_nontrivial R
+  · rw [isTranscendenceBasis_iff_of_subsingleton] at hx ⊢; infer_instance
+  rw [(hx.1.sumElim_comp hy.1).isTranscendenceBasis_iff_isAlgebraic]
+  set Rx := adjoin R (range x)
+  let Rxy := adjoin Rx (range y)
+  rw [show adjoin R (range <| Sum.elim y (algebraMap S A ∘ x)) = Rxy.restrictScalars R by
+    rw [← adjoin_algebraMap_image_union_eq_adjoin_adjoin, Sum.elim_range, union_comm, range_comp]]
+  show Algebra.IsAlgebraic Rxy A
+  have := hx.1.algebraMap_injective.nontrivial
+  have := hy.1.algebraMap_injective.nontrivial
+  have := hy.isAlgebraic
+  set Sy := adjoin S (range y)
+  let _ : Algebra Rxy Sy := by
+    refine (Subalgebra.inclusion (T := Sy.restrictScalars Rx) <| adjoin_le ?_).toAlgebra
+    rintro _ ⟨i, rfl⟩; exact subset_adjoin (s := range y) ⟨i, rfl⟩
+  have : IsScalarTower Rxy Sy A := .of_algebraMap_eq fun ⟨a, _⟩ ↦ show a = _ from rfl
+  have : IsScalarTower Rx Rxy Sy := .of_algebraMap_eq fun ⟨a, _⟩ ↦ Subtype.ext rfl
+  have : Algebra.IsAlgebraic Rxy Sy := by
+    refine ⟨fun ⟨a, ha⟩ ↦ adjoin_induction ?_ (fun _ ↦ .extendScalars (R := Rx) ?_ ?_)
+      (fun _ _ _ _ ↦ IsAlgebraic.add) (fun _ _ _ _ ↦ IsAlgebraic.mul) ha⟩
+    · rintro _ ⟨i, rfl⟩; exact isAlgebraic_algebraMap (⟨y i, subset_adjoin ⟨i, rfl⟩⟩ : Rxy)
+    · exact fun _ _ ↦ (Subtype.ext <| hy.1.algebraMap_injective <| Subtype.ext_iff.mp ·)
+    · exact (hx.isAlgebraic.1 _).algHom (IsScalarTower.toAlgHom Rx S Sy)
+  exact .trans' _ (S := Sy) Subtype.val_injective
+
+/-- If `s ⊆ t` are subsets in an `R`-algebra `A` such that `s` is algebraically independent over
+`R`, and `A` is algebraic over the `R`-algebra generated by `t`, then there is a transcendence
+basis of `A` over `R` between `s` and `t`, provided that `A` is a domain.
+
+This may fail if only `R` is assumed to be a domain but `A` is not, because of failure of
+transitivity of algebraicity: there may exist `a : A` such that `S := R[a]` is algebraic over
+`R` and `A` is algebraic over `S`, but `A` nonetheless contains a transcendental element over `R`.
+The only `R`-algebraically independent subset of `{a}` is `∅`, which is not a transcendence basis.
+See the docstring of `IsAlgebraic.restrictScalars_of_isIntegral` for an example. -/
+theorem exists_isTranscendenceBasis_between [NoZeroDivisors A] (s t : Set A) (hst : s ⊆ t)
```

Comment:
This looks like `Algebra.IsAlgebraic (adjoin R t) A` might be equivalent to `(AlgebraicIndependent.matroid _).Spanning t`. If so, the API for `Spanning` could be helpful going forward.

Reply:
Indeed, that's very satisfying! Might be able to golf a proof with it. Also I think `closure` is equal to `Subalgebra.algebraicClosure` of `Algebra.adjoin`.

Reply:
Possibly the way to organize things if you want to go this way is to have a file `Data.Matroid.Algebraic` that has the definition, as well as a bunch of lemmas like `spanning_iff`, `base_iff`, `flat_iff`, etc that relate the transcendence basis concepts to the matroid concepts. 

Then proofs for transcendence bases of matroidal facts will follow from rewrites and `Matroid` lemmas. For example, the above lemma is just `Indep.exists_base_subset_spanning` from #20885